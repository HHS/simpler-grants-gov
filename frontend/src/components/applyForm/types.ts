import {
  EnumOptionsType,
  FormContextType,
  GenericObjectType,
  RJSFSchema,
  StrictRJSFSchema,
  UIOptionsType,
} from "@rjsf/utils";

import { HTMLAttributes } from "react";

export interface FormData {
  [key: string]: FormDataEntryValue;
}

export interface SetFormDataFunction {
  (data: FormData): void;
}

export interface UiSchemaField {
  type: "field";
  definition: `/properties/${string}`;
}

export interface UiSchemaSection {
  type: "section";
  label: string;
  name: string;
  number?: string;
  children: (UiSchemaField | UiSchemaSection)[];
}

export type UiSchema =
  | (UiSchemaField | UiSchemaSection)
  | (UiSchemaField | UiSchemaSection)[];

export type SchemaField = {
  type: string;
  title: string;
  minLength?: number;
  maxLength?: number;
  format?: string;
};

export interface FormSchema {
  $schema: string;
  type: string;
  properties: Record<string, SchemaField>;
  required?: string[];
}

export type TextTypes =
  | "text"
  | "email"
  | "number"
  | "password"
  | "search"
  | "tel"
  | "url";

// extends the WidgetProps type from rjsf for USWDS and this project implementation
export interface UswdsWidgetProps<
  T = unknown,
  S extends StrictRJSFSchema = RJSFSchema,
  F extends FormContextType = never,
> extends GenericObjectType,
    Pick<
      HTMLAttributes<HTMLElement>,
      Exclude<keyof HTMLAttributes<HTMLElement>, "onBlur" | "onFocus">
    > {
  id: string;
  value?: string | unknown;
  type?: string;
  minLength?: number;
  schema: RJSFSchema;
  maxLength?: number;
  required?: boolean;
  disabled?: boolean;
  readonly?: boolean;
  hideError?: boolean;
  autofocus?: boolean;
  placeholder?: string;
  options: NonNullable<UIOptionsType<T, S, F>> & {
    /** The enum options list for a type that supports them */
    enumOptions?: EnumOptionsType<S>[];
    enumDisabled?: unknown;
    emptyValue?: string;
  };
  hideLabel?: boolean;
  multiple?: boolean;
  rawErrors?: string[] | undefined;
  // whether or not to use value + onChange
  updateOnInput?: boolean;
  onChange?: (value: unknown) => void;
  onBlur?: (id: string, value: unknown) => void;
  onFocus?: (id: string, value: unknown) => void;
}
