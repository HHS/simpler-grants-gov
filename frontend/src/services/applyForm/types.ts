export interface FormData {
  [key: string]: FormDataEntryValue;
}

export interface SetFormDataFunction {
  (data: FormData): void;
}

export interface UiSchemaField {
  type: "field";
  definition: string;
}

export interface UiSchemaSection {
  type: "section";
  label: string;
  name: string;
  number: string;
  children: (UiSchemaField | UiSchemaSection)[];
}

export type UiSchema =
  | (UiSchemaField | UiSchemaSection)
  | (UiSchemaField | UiSchemaSection)[];

export type SchemaField = {
  type: string;
  title: string;
  minLength?: number;
  maxLength?: number;
  format?: string;
};

export interface FormSchema {
  title: string;
  description?: string;
  properties: [SchemaField];
  required: [string];
}

export type TextTypes =
  | "text"
  | "email"
  | "number"
  | "password"
  | "search"
  | "tel"
  | "url";
