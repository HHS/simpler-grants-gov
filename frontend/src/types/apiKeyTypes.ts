export interface ApiKey {
  api_key_id: string;
  key_name: string;
  key_id: string;
  created_at: string;
  last_used: string | null;
  is_active: boolean;
}
