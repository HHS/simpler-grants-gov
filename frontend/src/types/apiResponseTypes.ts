import { QueryParamData } from "src/types/search/searchRequestTypes";

export interface PaginationInfo {
  order_by: string;
  page_offset: number;
  page_size: number;
  sort_direction: string;
  total_pages: number;
  total_records: number;
}

export interface APIResponse {
  data: unknown[] | object;
  message: string;
  status_code: number;
  pagination_info?: PaginationInfo;
  warnings?: unknown[] | null | undefined;
  errors?: unknown[] | null | undefined;
}

export interface FrontendErrorDetails {
  status?: number;
  searchInputs?: QueryParamData;
  field?: string;
  message?: string;
  type?: string;
}
