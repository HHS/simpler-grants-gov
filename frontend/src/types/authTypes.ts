import { JWTPayload } from "jose";

/**
 * Configure the UserProvider component.
 *
 * If you have any server-side rendered pages (using `getServerSideProps` or Server Components), you should get the
 * user from the server-side session and pass it to the `<UserProvider>` component via the `user`
 * prop. This will prefill the useUser hook with the UserProfile object.
 * For example:
 *
 * import { UserProvider } from 'src/services/auth/UserProvider';
 *
 * export default async function RootLayout({ children }) {
 *   // this will emit a warning because Server Components cannot write to cookies
 *   // see https://github.com/auth0/nextjs-auth0#using-this-sdk-with-react-server-components
 *   const session = await getSession();
 *
 *   return (
 *     <html lang="en">
 *       <body>
 *         <UserProvider user={session?.user}>
 *           {children}
 *         </UserProvider>
 *       </body>
 *     </html>
 *   );
 * }
 * ```
 *
 * In client-side rendered pages, the useUser hook uses a UserFetcher to fetch the
 * user from the profile API route. If needed, you can specify a custom fetcher here in the
 * `fetcher` option.
 *
 *
 * @category Client
 */

// represents relevant client side data from API JWT
export interface UserProfile {
  email?: string;
  token: string;
  expiresAt: Date;
  user_id: string;
}

// represents client JWT payload
export interface SimplerJwtPayload extends JWTPayload {
  token: string;
  session_duration_minutes: number;
}
// represents API JWT payload
export type UserSession = UserProfile & SimplerJwtPayload;

/**
 * Fetches the user from the profile API route to fill the useUser hook with the
 * UserProfile object.
 */
export type UserFetcher = (url: string) => Promise<UserSession | undefined>;

/**
 * @ignore
 */
export type UserProviderState = {
  user?: UserProfile;
  error?: Error;
  isLoading: boolean;
  refreshUser: () => Promise<void>;
};
