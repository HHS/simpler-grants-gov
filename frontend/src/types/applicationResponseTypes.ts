import { APIResponse } from "src/types/apiResponseTypes";

export interface ApplicationResponseDetail {
  [key: string]: string;
}

export interface ApplicationFormDetail {
  application_form_id: string;
  application_id: string;
  application_response: ApplicationResponseDetail;
  form_id: string;
}

export interface ApplicationDetail {
  application_forms: Array<ApplicationFormDetail>;
  application_id: string;
  competition_id: string;
}

export interface ApplicationStartApiResponse extends APIResponse {
  data: {
    application_id: string;
  };
}

export interface ApplicationFormDetailApiResponse extends APIResponse {
  data: ApplicationFormDetail;
}

export interface ApplicationDetailApiResponse extends APIResponse {
  data: ApplicationDetail;
}
