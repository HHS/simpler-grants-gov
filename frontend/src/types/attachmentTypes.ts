import { APIResponse } from "src/types/apiResponseTypes";

export type UploadStatus = "uploading" | "cancelled" | "completed" | "failed";

export type AttachmentUploadResponse = {
  application_attachment_id?: string;
};

export interface Attachment {
  application_attachment_id: string;
  download_path: string;
  created_at: string;
  file_name: string;
  file_size_bytes: number;
  mime_type: string;
  updated_at: string;
  status?: UploadStatus;
  cancelToken?: AbortController;
}
export interface AttachmentCardItem extends Attachment {
  id: string;
  file: File;
  status: UploadStatus;
  error?: string;
  abortController?: AbortController;
}

export interface AttachmentResponse extends APIResponse {
  data: Pick<
    Attachment,
    | "application_attachment_id"
    | "created_at"
    | "file_name"
    | "file_size_bytes"
    | "mime_type"
    | "updated_at"
    | "download_path"
  >[];
}
