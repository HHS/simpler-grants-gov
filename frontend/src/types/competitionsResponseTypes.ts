import { APIResponse } from "src/types/apiResponseTypes";

import { FormDetail } from "./formResponseTypes";

export interface CompetitionsDetailApiResponse extends APIResponse {
  data: {
    competition_forms: [{ form: FormDetail }];
    competition_id: string;
    opportunity_id: number;
  };
}
