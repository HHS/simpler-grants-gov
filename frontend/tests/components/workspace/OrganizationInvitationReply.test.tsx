import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe } from "jest-axe";
import { OrganizationInvitation } from "src/types/userTypes";
import { fakeOrganizationInvitation } from "src/utils/testing/fixtures";
import { useTranslationsMock } from "src/utils/testing/intlMocks";

import {
  InvitationAcceptedNotice,
  InvitationRejectedNotice,
  InvitationRejectionConfirmation,
  InvitationReplyForm,
  OrganizationInvitationReply,
} from "src/components/workspace/OrganizationInvitationReply";

const clientFetchMock = jest.fn();

jest.mock("next-intl", () => ({
  useTranslations: () => useTranslationsMock(),
}));

jest.mock("src/hooks/useClientFetch", () => ({
  useClientFetch: () => ({
    clientFetch: (...args: unknown[]) => clientFetchMock(...args) as unknown,
  }),
}));

const makeInvitation = (overrides: Partial<OrganizationInvitation>) => ({
  ...fakeOrganizationInvitation,
  ...overrides,
});

describe("OrganizationInvitationReply", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    clientFetchMock.mockResolvedValue({ status: "accepted" });
  });

  it("has no basic accessibility violations", async () => {
    const { container } = render(
      <OrganizationInvitationReply
        userInvitation={fakeOrganizationInvitation}
      />,
    );
    const results = await axe(container, {
      rules: {
        listitem: {
          enabled: false, // disabling this rule since in practice this component will always be part of a list
        },
      },
    });
    expect(results).toHaveNoViolations();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <OrganizationInvitationReply
        userInvitation={fakeOrganizationInvitation}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it("can fully accept a pending invitation", async () => {
    render(
      <OrganizationInvitationReply
        userInvitation={fakeOrganizationInvitation}
      />,
    );

    const acceptButton = screen.getByRole("button", { name: "accept" });
    const rejectButton = screen.getByRole("button", { name: "reject" });

    // no error
    expect(screen.queryByTestId("simpler-alert")).not.toBeInTheDocument();
    expect(acceptButton).toBeInTheDocument();
    expect(rejectButton).toBeInTheDocument();

    await userEvent.click(acceptButton);

    expect(clientFetchMock).toHaveBeenCalledWith(
      `/api/user/organization-invitations/${fakeOrganizationInvitation.organization_invitation_id}`,
      {
        method: "POST",
        body: JSON.stringify({
          accepted: true,
        }),
      },
    );

    expect(
      screen.getByRole("heading", { name: "accepted.ctaTitle" }),
    ).toBeInTheDocument();
  });

  it("can fully reject a pending invitation", async () => {
    clientFetchMock.mockResolvedValue({ status: "rejected" });
    render(
      <OrganizationInvitationReply
        userInvitation={fakeOrganizationInvitation}
      />,
    );

    const rejectButton = screen.getByRole("button", { name: "reject" });

    await userEvent.click(rejectButton);

    expect(clientFetchMock).not.toHaveBeenCalled();

    const confirmButton = screen.getByRole("button", {
      name: "rejectConfirmation.confirm",
    });
    const cancelButton = screen.getByRole("button", {
      name: "rejectConfirmation.cancel",
    });

    expect(confirmButton).toBeInTheDocument();
    expect(cancelButton).toBeInTheDocument();

    await userEvent.click(confirmButton);

    expect(clientFetchMock).toHaveBeenCalledWith(
      `/api/user/organization-invitations/${fakeOrganizationInvitation.organization_invitation_id}`,
      {
        method: "POST",
        body: JSON.stringify({
          accepted: false,
        }),
      },
    );

    expect(
      screen.getByRole("heading", { name: "rejected.ctaTitle" }),
    ).toBeInTheDocument();
  });
});

describe("InvitationReplyForm", () => {
  it("displays an error when passed", async () => {
    const mockOnErrorClick = jest.fn();
    render(
      <InvitationReplyForm
        error={new Error()}
        onErrorClick={mockOnErrorClick}
        organizationName={"hi"}
        organizationInvitationId={"id"}
        onAccept={jest.fn()}
        onReject={jest.fn()}
      />,
    );

    expect(screen.getByTestId("simpler-alert")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <InvitationReplyForm
        onErrorClick={jest.fn()}
        organizationName={"hi"}
        organizationInvitationId={"id"}
        onAccept={jest.fn()}
        onReject={jest.fn()}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe("InvitationAcceptedNotice", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <InvitationAcceptedNotice organizationName="hi" />,
    );
    expect(container).toMatchSnapshot();
  });
});

describe("InvitationRejectedNotice", () => {
  it("matches snapshot", () => {
    const { container } = render(<InvitationRejectedNotice />);
    expect(container).toMatchSnapshot();
  });
});

describe("InvitationRejectionConfirmation", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <InvitationRejectionConfirmation
        onCancel={jest.fn()}
        onConfirm={jest.fn()}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
