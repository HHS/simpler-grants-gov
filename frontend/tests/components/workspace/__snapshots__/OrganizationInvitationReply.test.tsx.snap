// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvitationAcceptedNotice matches snapshot 1`] = `
<div>
  <li
    class="border-primary border-2px padding-2 margin-top-2 radius-md"
  >
    <div
      class="grid-row"
      data-testid="grid"
    >
      <div
        class="tablet:grid-col-8"
        data-testid="grid"
      >
        <h3>
          accepted.ctaTitle
        </h3>
        <p>
          accepted.description
        </p>
      </div>
    </div>
  </li>
</div>
`;

exports[`InvitationRejectedNotice matches snapshot 1`] = `
<div>
  <li
    class="border-primary border-2px padding-2 margin-top-2 radius-md"
  >
    <div
      class="grid-row"
      data-testid="grid"
    >
      <div
        class="tablet:grid-col-8"
        data-testid="grid"
      >
        <h3>
          rejected.ctaTitle
        </h3>
        <p>
          rejected.description
        </p>
      </div>
    </div>
  </li>
</div>
`;

exports[`InvitationRejectionConfirmation matches snapshot 1`] = `
<div>
  <li
    class="border-primary border-2px padding-2 margin-top-2 radius-md"
  >
    <div
      class="grid-row"
      data-testid="grid"
    >
      <div
        class="tablet:grid-col-8"
        data-testid="grid"
      >
        <h3>
          rejectConfirmation.ctaTitle
        </h3>
        <p>
          rejectConfirmation.description
        </p>
      </div>
      <div
        class="tablet:grid-col-4 text-right flex-align-self-center"
        data-testid="grid"
      >
        <button
          class="usa-button usa-button--accent-warm margin-left-2"
          data-testid="button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="usa-icon usa-icon usa-icon--size-4"
            role="img"
          >
            <use
              href="/img.jpg#do_not_disturb"
            />
          </svg>
          rejectConfirmation.confirm
        </button>
        <button
          class="usa-button usa-button--unstyled margin-left-2"
          data-testid="button"
          type="button"
        >
          rejectConfirmation.cancel
        </button>
      </div>
    </div>
  </li>
</div>
`;

exports[`InvitationReplyForm matches snapshot 1`] = `
<div>
  <li
    class="border-primary border-2px padding-2 margin-top-2 radius-md"
  >
    <div
      class="grid-row"
      data-testid="grid"
    >
      <div
        class="tablet:grid-col-8"
        data-testid="grid"
      >
        <h3>
          hi
           
          ctaTitle
        </h3>
        <p>
          description
        </p>
      </div>
      <div
        class="tablet:grid-col-4 text-right flex-align-self-center"
        data-testid="grid"
      >
        <button
          class="usa-button margin-left-2"
          data-testid="button"
          type="button"
        >
          accept
        </button>
        <button
          class="usa-button usa-button--secondary margin-left-2"
          data-testid="button"
          type="button"
        >
          reject
        </button>
      </div>
    </div>
  </li>
</div>
`;

exports[`OrganizationInvitationReply matches snapshot 1`] = `
<div>
  <li
    class="border-primary border-2px padding-2 margin-top-2 radius-md"
  >
    <div
      class="grid-row"
      data-testid="grid"
    >
      <div
        class="tablet:grid-col-8"
        data-testid="grid"
      >
        <h3>
          Example Organization
           
          ctaTitle
        </h3>
        <p>
          description
        </p>
      </div>
      <div
        class="tablet:grid-col-4 text-right flex-align-self-center"
        data-testid="grid"
      >
        <button
          class="usa-button margin-left-2"
          data-testid="button"
          type="button"
        >
          accept
        </button>
        <button
          class="usa-button usa-button--secondary margin-left-2"
          data-testid="button"
          type="button"
        >
          reject
        </button>
      </div>
    </div>
  </li>
</div>
`;
